enumerable_methods.rb

```diff
@@ -0,0 +1,180 @@
+
+  def my_select
+    return enum_for(:my_select) unless block_given?
+
+    arr = *self
+    new_arr = []
+    arr.my_each do |element|
+      next unless yield(element)
+
+      new_arr.push(element)
+    end
+    new_arr
+  end
+
+  # my_all?
+
+  def my_all?(arg = nil)
+    condition = true
+    my_each do |element|
+      if arg.is_a?(Class)
+        if element.is_a?(arg) == false
+          condition = false
+          break
+        end
+      elsif arg.is_a?(Regexp)
+        if element.match?(arg) == false
+          condition = false
+          break
+        end
+      else
+        if block_given?
+          if yield(element) == false || element.nil?
+            condition = false
+            break
+          end
+        else
+          if element == false || element.nil?
+            condition = false
+            break
+          end
+        end
+      end
+    end
+    condition
+  end
+
+  # my_any?
+
+  def my_any?(arg = nil)
+    condition = false
+    my_each do |element|
+      if arg.is_a?(Class)
+        if element.is_a?(arg)
+          condition = true
+          break
+        end
+      elsif arg.is_a?(Regexp)
+        if element.match?(arg)
+          condition = true
+          break
+        end
+      else
+        if block_given?
+          if yield(element) || element.nil?
+            condition = true
+            break
+          end
+        else
+          if element || element.nil?
+            condition = true
+            break
+          end
+        end
+      end
+    end
+    condition
+  end
+
+  # my_none?
+
+  def my_none?(arg = nil)
+    !my_any?(arg)
+  end
+
+  # count
+
+  def my_count(arg = nil)
+    arr = *self
+    if block_given? && arg.nil?
+      new_arr = []
+      arr.my_each do |element|
+        next unless yield(element)
+
+        new_arr.push(element)
+      end
+      new_arr.length
+    elsif (!block_given? && !arg.nil?) || (block_given? && !arg.nil?)
+      new_arr = []
+      arr.my_each do |element|
+        new_arr.push(element) if element == arg
+      end
+      new_arr.length
+    else
+      arr.length
+    end
+  end
+
+  # my_map
+
+  def my_map(my_proc = nil)
+    return enum_for(:my_map) unless block_given?
+
+    new_arr = []
+    if my_proc
+      my_each { |i| new_arr.push my_proc.call(i) }
+    else
+      my_each { |i| new_arr.push(yield(i)) }
+    end
+    new_arr
+  end
+
+  # my_inject
+
+  def my_inject(arg1 = 0, arg2 = nil)
+    return raise LocalJumpError if !block_given? && !arg1
+
+    if block_given?
+      result = arg1
+      my_each do |element|
+        result = yield(result, element)
+      end
+    elsif !block_given? && arg1 && arg2
+      result = arg1
+      my_each do |element|
+        result = result.send arg2, element
+      end
+    elsif !block_given? && !arg2
+      result = 0
+      result = 1 if arg1 == (:* || '*')
+      my_each do |element|
+        result = result.send arg1, element
+      end
+    end
+    result
+  end
+end
```

Comment:
- [ ] `my_inject` should raise a `localJumpError` when no block or argument is given. 
    - Yours crrently retrieves a `typeError`
```ruby
array = [1,2,3,4,5]
array.my_inject
```
screenshot of the `my_inject` error message 👇 
![Screen Shot 2020-09-22 at 8 08 31 PM](https://user-images.githubusercontent.com/48804258/93920376-663f9080-fd0f-11ea-993e-db2e9aa1c89d.png)

```ruby
array = [1,2,3,4,5]
array.inject
```
screenshot of the `inject` error message 👇 
![Screen Shot 2020-09-22 at 8 10 08 PM](https://user-images.githubusercontent.com/48804258/93920551-9f780080-fd0f-11ea-87d0-6fc2e2291d15.png)
****

- [ ] `my_inject` should search for the longest word in an array of strings.
 ```ruby
search =  proc { |memo, word| memo.length > word.length ? memo : word } 
words = %w[dog door rod blade]
words.my_inject(&search) 
```
- When trying to test this, your code throws an error 👇 
![Screen Shot 2020-09-22 at 8 16 37 PM](https://user-images.githubusercontent.com/48804258/93921198-86238400-fd10-11ea-9337-9dde4ca00ce8.png)
results from the real `inject` method 👇 
![Screen Shot 2020-09-22 at 8 16 56 PM](https://user-images.githubusercontent.com/48804258/93921226-9176af80-fd10-11ea-9742-696e8bd44bf0.png)
****

- [ ] When `my_inject` is called on a range and given a block without an initial value, it should combine all the elements of enum by applying a binary operation.
```ruby
 range = Range.new(5, 50)
range.my_inject { |prod, n| prod * n } === range.inject { |prod, n| prod * n } => false
```
_Hint: Yours returns 0_